vmm/src/cpu.rs

```diff
@@ -1018,6 +1025,25 @@ impl CpuManager {
                                 vcpu_run_interrupted.store(false, Ordering::SeqCst);
                             }
 
+                            if vcpu_kick_signalled.load(Ordering::SeqCst) {
+                                #[cfg(feature = "kvm")]
+                                if matches!(hypervisor_type, HypervisorType::Kvm) {
+                                    vcpu.lock().as_ref().unwrap().vcpu.set_immediate_exit(true);
+                                    if !matches!(vcpu.lock().unwrap().run(), Ok(VmExit::Ignore)) {
+                                        error!("Unexpected VM exit on \"immediate_exit\" run");
+                                        break;
+                                    }
+                                    vcpu.lock().as_ref().unwrap().vcpu.set_immediate_exit(false);
+                                }
```

Comment:
After thinking a bit more about it. Usually, you would inject an NMI when a CPU is not progressing to check where it is stuck. So, I don't think it makes sense to run the vcpu first to complete the pending IO because it would not finish instead directly inject the NMI to get a backtrace from that vcpu where it is stuck.

Reply:
Thx for the suggestion! I have reworked this part, and PTAL.